### ISSUE TITLE
OFS 3.0 doesn't close properly

### ISSUE THREAD

**Issue by Zalunda** [OpenFunscripter/OFS#61]
I started and closed a few OFS 3.0 and noticed they were still in the task manager.

I don't know if it can help but this is what "Process  Explorer" shows for the 3 processes in memory:

![image](https://user-images.githubusercontent.com/8295521/208265607-fd75cd02-429d-4b71-839b-84ba820f6344.png)

I'm using "OFS-64x_AVX_6dc9a778c0d2cf8b1e554b3615f00f9cd9955634.zip".

**Comment by OpenFunscripter** [comment]
That's odd this doesn't happen for me.
Is the CPU usage high?

**Comment by OpenFunscripter** [comment]
The fact that there's no mpv-2.dll in there is a strong indicator that it's stuck somwhere here.
Did you turn on the new websocket api?

https://github.com/OpenFunscripter/OFS/blob/b4fbe440e21554ddd7305c5daa6d186725e42aca/src/OpenFunscripter.cpp#L1711-L1716

**Comment by Zalunda** [comment]
CPU Usage: Not too bad but each instance used about 12% CPU. They were not completely idle. 
And no, I didn't turn on websocket api (unless it's turned on by default).

I did a few other tests, it's always occurring. I also get the same problem with the non-AVX version.

I'll try to compile / debug in visual studio to see if I can get more info.

Threads state while running vs stopped/bugged:
![image](https://user-images.githubusercontent.com/8295521/208268565-6256ecb4-522d-4b76-baa4-3623f05b688d.png)

![image](https://user-images.githubusercontent.com/8295521/208268575-b0c402f6-5e62-405c-a5ea-d2db631f1f7d.png)

**Comment by OpenFunscripter** [comment]
The websocket api is off by default but a thread is created nonetheless.

> CPU Usage: Not too bad but each instance used about 12% CPU. They were not completely idle.

Well that is pretty high because it's probably maxing out one core, spin waiting on something.
I just pushed a change a bit of a stab in the dark.

You don't need to compile it yourself you can download the artifact here https://github.com/OpenFunscripter/OFS/actions/runs/3722105687/jobs/6312697423 once it's finished.

**Comment by Zalunda** [comment]
Ok, it loop in WGI_JoystickQuit:
```
        while (wgi.controller_count > 0) {
            IEventHandler_CRawGameControllerVtbl_InvokeRemoved(&controller_removed.iface, NULL, wgi.controllers[wgi.controller_count - 1].controller);
        }
```
wgi.controller_count is always 1

![image](https://user-images.githubusercontent.com/8295521/208269238-5e41299a-4201-4f0f-819b-886d0298e9bb.png)

![image](https://user-images.githubusercontent.com/8295521/208269392-f89066f2-5db2-4fd5-947a-577335f8b152.png)

I'm not sure which joystick it is. I tried to unplug my Xbox controller and it still loop at the same place.

**Comment by OpenFunscripter** [comment]
Oh wow in this may not even be my fault.
Thank you for checking in a debugger 😄

So it's actually stuck deep in the `SDL_Quit()`
I did update to the latest SDL 2.26.1 for this release.

I will check if I'm doing everything in my power to close open controllers.

**Comment by Zalunda** [comment]
I downloaded and tested the previous version before the update. It's definitely the library update.

Previous: https://github.com/OpenFunscripter/OFS/actions/runs/3714873553  => WORK
After Updating libraries commit: https://github.com/OpenFunscripter/OFS/actions/runs/3720177007 => STUCK

**Comment by OpenFunscripter** [comment]
I added some code to close open controllers on shutdown on my end.
I don't think that this will fix the issue for you. Nonetheless it would be great if you could try this.

SDL2 has multiple drivers for input devices.
I have PS4 controller and it's using the HIDAPI driver and not the WGI one. I don't know if I can influence this.
So in my case when I step into `WGI_JoystickQuit` there's 0 controllers because my PS4 controller isn't handled by that driver.
If my change doesn't doesn't work for you I'll look into compiling SDL2 without `Windows.Gaming.Input` support.
I don't think it's needed

**Comment by Zalunda** [comment]
Yup, it didn't fix for me. If you could disable joysticks, it would be nice. I might not be the only one that has a WGI device.

But I will also open a bug on SDL2. 

Looking at the code, I'm not sure how a WGI joystick driver could be closed correctly.

My guess is that this commit broke something: 
https://github.com/libsdl-org/SDL/commit/40bd4feedcca779a4600743c27a786b436edc8f7#diff-1e92289df6a025c21fcb6e1af73e63bec8606093686f506aa78773e8409b94e9

It starts with:

Important line: **SDL_joysticks_quitting = SDL_TRUE;**
```
void
SDL_JoystickQuit(void)
{
    int i;

    SDL_LockJoysticks();

    SDL_joysticks_quitting = SDL_TRUE;

    /* Stop the event polling */
    while (SDL_joysticks) {
        SDL_joysticks->ref_count = 1;
        SDL_JoystickClose(SDL_joysticks);
    }

    /* Quit drivers in reverse order to avoid breaking dependencies between drivers */
    for (i = SDL_arraysize(SDL_joystick_drivers) - 1; i >= 0; --i) {
       SDL_joystick_drivers[i]->Quit();
    }
```
And then, for a WGI joystick driver:

```
static void
WGI_JoystickQuit(void)
{
    if (wgi.statics) {
        while (wgi.controller_count > 0) {
            IEventHandler_CRawGameControllerVtbl_InvokeRemoved(&controller_removed.iface, NULL, wgi.controllers[wgi.controller_count - 1].controller);
        }
     ...   
```
And then: 
```
static HRESULT STDMETHODCALLTYPE IEventHandler_CRawGameControllerVtbl_InvokeRemoved(__FIEventHandler_1_Windows__CGaming__CInput__CRawGameController * This, IInspectable *sender, __x_ABI_CWindows_CGaming_CInput_CIRawGameController *e)
{
    HRESULT hr;
    __x_ABI_CWindows_CGaming_CInput_CIRawGameController *controller = NULL;

    SDL_LockJoysticks();

    /* Can we get delayed calls to InvokeRemoved() after WGI_JoystickQuit()? */
    if (SDL_JoysticksQuitting() || !SDL_JoysticksInitialized()) {
        SDL_UnlockJoysticks();
        return S_OK;
    }
    
    ...
    --wgi.controller_count;
    ...
```

But it's impossible to get to the "--wgi.controller_count" line that would end the loop 
because SDL_JoysticksQuitting() is always true  (i.e. **SDL_joysticks_quitting = SDL_TRUE;** line from the start)

```
SDL_bool
SDL_JoysticksQuitting(void)
{
    return SDL_joysticks_quitting;
}
```

**Comment by Zalunda** [comment]
FYI: https://github.com/libsdl-org/SDL/issues/6837

**Comment by OpenFunscripter** [comment]
All right will keep an eye on this.
I suppose rather than compiling without WGI I just rollback to the previous version.
I didn't have a reason to update I just like to update 🙃
